File: 2021/Day 6/Javascript/code.js
// Thank you! Alan Malloy! For the concept!
exports.part1 = (input) => {
    let ageMap = new Map();

    input.split(',').map(Number).forEach((v) => {
        ageMap.set(v, ageMap.get(v) ? ageMap.get(v) + 1 : 1);
    });

    for (let day = 0; day < 80; day++) {
        const newAgeMap = new Map();
        for (let [key, value] of ageMap) {
            if (key == 0) {
                newAgeMap.set(6, (newAgeMap.get(6) || 0) + value);
                newAgeMap.set(8, (newAgeMap.get(8) || 0) + value);
            } else {
                newAgeMap.set(key - 1, (newAgeMap.get(key - 1) || 0) + value);
            }
        }
        ageMap = newAgeMap;
    }

    let fishCount = 0;

    for (let [_, value] of ageMap) {
        fishCount += value;
    }

    return fishCount;
}

exports.part2 = (input) => {
    let ageMap = new Map();

    input.split(',').map(Number).forEach((v) => {
        ageMap.set(v, ageMap.get(v) ? ageMap.get(v) + 1 : 1);
    });

    for (let day = 0; day < 256; day++) {
        const newAgeMap = new Map();
        for (let [key, value] of ageMap) {
            if (key == 0) {
                newAgeMap.set(6, (newAgeMap.get(6) || 0) + value);
                newAgeMap.set(8, (newAgeMap.get(8) || 0) + value);
            } else {
                newAgeMap.set(key - 1, (newAgeMap.get(key - 1) || 0) + value);
            }
        }
        ageMap = newAgeMap;
    }

    let fishCount = 0;

    for (let [_, value] of ageMap) {
        fishCount += value;
    }

    return fishCount;
}

File: 2021/Day 6/lua/code.lua
function dofile(filename)
	return assert(loadfile(filename))()
end

local Utility = dofile("..\\..\\Common\\lua\\Utility.lua")

local code = {}

function code:part1(input)
	local ageObject = {}

	Utility
		.split(input, ",", true)
		:map(function(x)
			return tonumber(x)
		end)
		:forEach(function(v)
			ageObject[v] = (ageObject[v] and ageObject[v] + 1) or 1
		end)

	for day = 1, 80 do
		local newAgeObject = {}
		for k, v in pairs(ageObject) do
			if k == 0 then
				newAgeObject[6] = (newAgeObject[6] or 0) + v
				newAgeObject[8] = (newAgeObject[8] or 0) + v
			else
				newAgeObject[k - 1] = (newAgeObject[k - 1] or 0) + v
			end
		end
		ageObject = newAgeObject
	end

	local fishCount = 0

	for _, v in pairs(ageObject) do
		fishCount = fishCount + v
	end

	return fishCount
end

function code:part2(input)
	local ageObject = {}

	Utility
		.split(input, ",", true)
		:map(function(x)
			return tonumber(x)
		end)
		:forEach(function(v)
			ageObject[v] = (ageObject[v] and ageObject[v] + 1) or 1
		end)

	for day = 1, 256 do
		local newAgeObject = {}
		for k, v in pairs(ageObject) do
			if k == 0 then
				newAgeObject[6] = (newAgeObject[6] or 0) + v
				newAgeObject[8] = (newAgeObject[8] or 0) + v
			else
				newAgeObject[k - 1] = (newAgeObject[k - 1] or 0) + v
			end
		end
		ageObject = newAgeObject
	end

	local fishCount = 0

	for _, v in pairs(ageObject) do
		fishCount = fishCount + v
	end

	return fishCount
end

return code
